In [13]:
# Importing Classifier Modules
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.ensemble import VotingClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from xgboost import XGBClassifier


import numpy as np
import pandas as pd
import matplotlib.pylab as plt

#import warnings
#warnings.filterwarnings('ignore')

In [14]:
# load data
train = pd.read_csv('./train.csv').iloc[:, 1:]
test = pd.read_csv('./test.csv')

# split data into X and y
dfX = train.iloc[:,:-1]
dfy = train.iloc[:,[-1]]

# 데이터 탐색

In [15]:
train.tail()
#test.tail()

Elevation  Aspect  Slope  Horizontal_Distance_To_Hydrology  \
15115       2607     243     23                               258   
15116       2603     121     19                               633   
15117       2492     134     25                               365   
15118       2487     167     28                               218   
15119       2475     197     34                               319   

       Vertical_Distance_To_Hydrology  Horizontal_Distance_To_Roadways  \
15115                               7                              660   
15116                             195                              618   
15117                             117                              335   
15118                             101                              242   
15119                              78                              270   

       Hillshade_9am  Hillshade_Noon  Hillshade_3pm  \
15115            170             251            214   
15116            249             221             91   
15117            250             220             83   
15118            229             237            119   
15119            189             244            164   

       Horizontal_Distance_To_Fire_Points     ...      Soil_Type32  \
15115                                1282     ...                0   
15116                                1325     ...                0   
15117                                1187     ...                0   
15118                                 932     ...                0   
15119                                 914     ...                0   

       Soil_Type33  Soil_Type34  Soil_Type35  Soil_Type36  Soil_Type37  \
15115            0            0            0            0            0   
15116            0            0            0            0            0   
15117            0            0            0            0            0   
15118            0            0            0            0            0   
15119            0            0            0            0            0   

       Soil_Type38  Soil_Type39  Soil_Type40  Cover_Type  
15115            0            0            0           3  
15116            0            0            0           3  
15117            0            0            0           3  
15118            0            0            0           3  
15119            0            0            0           3  

[5 rows x 55 columns]

In [16]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15120 entries, 0 to 15119
Data columns (total 55 columns):
Elevation                             15120 non-null int64
Aspect                                15120 non-null int64
Slope                                 15120 non-null int64
Horizontal_Distance_To_Hydrology      15120 non-null int64
Vertical_Distance_To_Hydrology        15120 non-null int64
Horizontal_Distance_To_Roadways       15120 non-null int64
Hillshade_9am                         15120 non-null int64
Hillshade_Noon                        15120 non-null int64
Hillshade_3pm                         15120 non-null int64
Horizontal_Distance_To_Fire_Points    15120 non-null int64
Wilderness_Area1                      15120 non-null int64
Wilderness_Area2                      15120 non-null int64
Wilderness_Area3                      15120 non-null int64
Wilderness_Area4                      15120 non-null int64
Soil_Type1                            15120 non-null int64
Soil_T

In [17]:
train.groupby('Cover_Type').size()

Cover_Type
1    2160
2    2160
3    2160
4    2160
5    2160
6    2160
7    2160
dtype: int64

In [18]:
train.iloc[:,:20].describe()

Elevation        Aspect         Slope  \
count  15120.000000  15120.000000  15120.000000   
mean    2749.322553    156.676653     16.501587   
std      417.678187    110.085801      8.453927   
min     1863.000000      0.000000      0.000000   
25%     2376.000000     65.000000     10.000000   
50%     2752.000000    126.000000     15.000000   
75%     3104.000000    261.000000     22.000000   
max     3849.000000    360.000000     52.000000   

       Horizontal_Distance_To_Hydrology  Vertical_Distance_To_Hydrology  \
count                      15120.000000                    15120.000000   
mean                         227.195701                       51.076521   
std                          210.075296                       61.239406   
min                            0.000000                     -146.000000   
25%                           67.000000                        5.000000   
50%                          180.000000                       32.000000   
75%                          330.000000                       79.000000   
max                         1343.000000                      554.000000   

       Horizontal_Distance_To_Roadways  Hillshade_9am  Hillshade_Noon  \
count                     15120.000000   15120.000000    15120.000000   
mean                       1714.023214     212.704299      218.965608   
std                        1325.066358      30.561287       22.801966   
min                           0.000000       0.000000       99.000000   
25%                         764.000000     196.000000      207.000000   
50%                        1316.000000     220.000000      223.000000   
75%                        2270.000000     235.000000      235.000000   
max                        6890.000000     254.000000      254.000000   

       Hillshade_3pm  Horizontal_Distance_To_Fire_Points  Wilderness_Area1  \
count   15120.000000                        15120.000000      15120.000000   
mean      135.091997                         1511.147288          0.237897   
std        45.895189                         1099.936493          0.425810   
min         0.000000                            0.000000          0.000000   
25%       106.000000                          730.000000          0.000000   
50%       138.000000                         1256.000000          0.000000   
75%       167.000000                         1988.250000          0.000000   
max       248.000000                         6993.000000          1.000000   

       Wilderness_Area2  Wilderness_Area3  Wilderness_Area4    Soil_Type1  \
count      15120.000000      15120.000000      15120.000000  15120.000000   
mean           0.033003          0.419907          0.309193      0.023479   
std            0.178649          0.493560          0.462176      0.151424   
min            0.000000          0.000000          0.000000      0.000000   
25%            0.000000          0.000000          0.000000      0.000000   
50%            0.000000          0.000000          0.000000      0.000000   
75%            0.000000          1.000000          1.000000      0.000000   
max            1.000000          1.000000          1.000000      1.000000   

         Soil_Type2    Soil_Type3    Soil_Type4    Soil_Type5    Soil_Type6  
count  15120.000000  15120.000000  15120.000000  15120.000000  15120.000000  
mean       0.041204      0.063624      0.055754      0.010913      0.042989  
std        0.198768      0.244091      0.229454      0.103896      0.202840  
min        0.000000      0.000000      0.000000      0.000000      0.000000  
25%        0.000000      0.000000      0.000000      0.000000      0.000000  
50%        0.000000      0.000000      0.000000      0.000000      0.000000  
75%        0.000000      0.000000      0.000000      0.000000      0.000000  
max        1.000000      1.000000      1.000000      1.000000      1.000000

In [19]:
train.iloc[:,20:40].describe()

Soil_Type7    Soil_Type8    Soil_Type9   Soil_Type10   Soil_Type11  \
count     15120.0  15120.000000  15120.000000  15120.000000  15120.000000   
mean          0.0      0.000066      0.000661      0.141667      0.026852   
std           0.0      0.008133      0.025710      0.348719      0.161656   
min           0.0      0.000000      0.000000      0.000000      0.000000   
25%           0.0      0.000000      0.000000      0.000000      0.000000   
50%           0.0      0.000000      0.000000      0.000000      0.000000   
75%           0.0      0.000000      0.000000      0.000000      0.000000   
max           0.0      1.000000      1.000000      1.000000      1.000000   

        Soil_Type12   Soil_Type13   Soil_Type14  Soil_Type15   Soil_Type16  \
count  15120.000000  15120.000000  15120.000000      15120.0  15120.000000   
mean       0.015013      0.031481      0.011177          0.0      0.007540   
std        0.121609      0.174621      0.105133          0.0      0.086506   
min        0.000000      0.000000      0.000000          0.0      0.000000   
25%        0.000000      0.000000      0.000000          0.0      0.000000   
50%        0.000000      0.000000      0.000000          0.0      0.000000   
75%        0.000000      0.000000      0.000000          0.0      0.000000   
max        1.000000      1.000000      1.000000          0.0      1.000000   

        Soil_Type17   Soil_Type18   Soil_Type19   Soil_Type20   Soil_Type21  \
count  15120.000000  15120.000000  15120.000000  15120.000000  15120.000000   
mean       0.040476      0.003968      0.003042      0.009193      0.001058   
std        0.197080      0.062871      0.055075      0.095442      0.032514   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

        Soil_Type22   Soil_Type23   Soil_Type24   Soil_Type25   Soil_Type26  
count  15120.000000  15120.000000  15120.000000  15120.000000  15120.000000  
mean       0.022817      0.050066      0.016997      0.000066      0.003571  
std        0.149326      0.218089      0.129265      0.008133      0.059657  
min        0.000000      0.000000      0.000000      0.000000      0.000000  
25%        0.000000      0.000000      0.000000      0.000000      0.000000  
50%        0.000000      0.000000      0.000000      0.000000      0.000000  
75%        0.000000      0.000000      0.000000      0.000000      0.000000  
max        1.000000      1.000000      1.000000      1.000000      1.000000

In [20]:
train.iloc[:,40:55].describe()

Soil_Type27   Soil_Type28   Soil_Type29   Soil_Type30   Soil_Type31  \
count  15120.000000  15120.000000  15120.000000  15120.000000  15120.000000   
mean       0.000992      0.000595      0.085384      0.047950      0.021958   
std        0.031482      0.024391      0.279461      0.213667      0.146550   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

        Soil_Type32   Soil_Type33   Soil_Type34   Soil_Type35   Soil_Type36  \
count  15120.000000  15120.000000  15120.000000  15120.000000  15120.000000   
mean       0.045635      0.040741      0.001455      0.006746      0.000661   
std        0.208699      0.197696      0.038118      0.081859      0.025710   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

        Soil_Type37   Soil_Type38   Soil_Type39   Soil_Type40    Cover_Type  
count  15120.000000  15120.000000  15120.000000  15120.000000  15120.000000  
mean       0.002249      0.048148      0.043452      0.030357      4.000000  
std        0.047368      0.214086      0.203880      0.171574      2.000066  
min        0.000000      0.000000      0.000000      0.000000      1.000000  
25%        0.000000      0.000000      0.000000      0.000000      2.000000  
50%        0.000000      0.000000      0.000000      0.000000      4.000000  
75%        0.000000      0.000000      0.000000      0.000000      6.000000  
max        1.000000      1.000000      1.000000      1.000000      7.000000

* train datasets 15119개
* test datasets 565891개
* 독립변수 54개(더미변수 형태 제외 12개)
* 종속변수 1개 Cover_type
* 클래스 별로 일정한 수를 가지고 있음
* 'Soil_Type7', 'Soil_Type15'는 데이터가 0

# 변수 설명

* Elevation - Elevation in meters (높이 /meter 단위)
* Aspect - Aspect in degrees azimuth (방위각)
* Slope - Slope in degrees (기울기 각도 /도 단위)
* Horizontal_Distance_To_Hydrology - Horz Dist to nearest surface water features (수원과의 수평거리)
* Vertical_Distance_To_Hydrology - Vert Dist to nearest surface water features (수원과의 수직거리)
* Horizontal_Distance_To_Roadways - Horz Dist to nearest roadway (길가와의 수평거리)
* Hillshade_9am (0 to 255 index) - Hillshade index at 9am, summer solstice (오전 9시의 차양 / 0~255)
* Hillshade_Noon (0 to 255 index) - Hillshade index at noon, summer solstice (정오시의 차양/ 0~255)
* Hillshade_3pm (0 to 255 index) - Hillshade index at 3pm, summer solstice (오후 9시의 차양/ 0~255)
* Horizontal_Distance_To_Fire_Points - Horz Dist to nearest wildfire ignition points (야생 산불 발화지점과의 수평거리)
* Wilderness_Area (4 binary columns, 0 = absence or 1 = presence) - Wilderness area designation (황야 지대 /4종류 ) in Roosevelt National Forest of northern Colorado
* Soil_Type (40 binary columns, 0 = absence or 1 = presence) - Soil Type designation (토양 종류 / 40종류)
* 토양종류와 황야 지대 카테고리별 설명은 https://www.kaggle.com/c/forest-cover-type-prediction/data 참조

# 종속 변수

* Cover_Type (7 types, integers 1 to 7) - Forest Cover Type designation- (산림 유형 / 7종류) (the predominant kind of tree cover)
* 1 - Spruce/Fir
* 2 - Lodgepole Pine
* 3 - Ponderosa Pine
* 4 - Cottonwood/Willow
* 5 - Aspen
* 6 - Douglas-fir
* 7 - Krummholz

# 비슷한 feature 결합

In [21]:
train['Distanse_to_Hydrolody'] = (train['Horizontal_Distance_To_Hydrology']**2+train['Vertical_Distance_To_Hydrology']**2)**0.5
test['Distanse_to_Hydrolody'] = (test['Horizontal_Distance_To_Hydrology']**2+test['Vertical_Distance_To_Hydrology']**2)**0.5

train['Hydro_Fire_1'] = train['Horizontal_Distance_To_Hydrology']+train['Horizontal_Distance_To_Fire_Points']
test['Hydro_Fire_1'] = test['Horizontal_Distance_To_Hydrology']+test['Horizontal_Distance_To_Fire_Points']

train['Hydro_Fire_2'] = abs(train['Horizontal_Distance_To_Hydrology']-train['Horizontal_Distance_To_Fire_Points'])
test['Hydro_Fire_2'] = abs(test['Horizontal_Distance_To_Hydrology']-test['Horizontal_Distance_To_Fire_Points'])

train['Hydro_Road_1'] = abs(train['Horizontal_Distance_To_Hydrology']+train['Horizontal_Distance_To_Roadways'])
test['Hydro_Road_1'] = abs(test['Horizontal_Distance_To_Hydrology']+test['Horizontal_Distance_To_Roadways'])

train['Hydro_Road_2'] = abs(train['Horizontal_Distance_To_Hydrology']-train['Horizontal_Distance_To_Roadways'])
test['Hydro_Road_2'] = abs(test['Horizontal_Distance_To_Hydrology']-test['Horizontal_Distance_To_Roadways'])

train['Fire_Road_1'] = abs(train['Horizontal_Distance_To_Fire_Points']+train['Horizontal_Distance_To_Roadways'])
test['Fire_Road_1'] = abs(test['Horizontal_Distance_To_Fire_Points']+test['Horizontal_Distance_To_Roadways'])

train['Fire_Road_2'] = abs(train['Horizontal_Distance_To_Fire_Points']-train['Horizontal_Distance_To_Roadways'])
test['Fire_Road_2'] = abs(test['Horizontal_Distance_To_Fire_Points']-test['Horizontal_Distance_To_Roadways'])

# Data Handling

In [22]:
# Delete empty colums
# From both train and test data
train = train.drop(['Soil_Type7', 'Soil_Type15'], axis = 1)
test = test.drop(['Soil_Type7', 'Soil_Type15'], axis = 1)

# split data into X and y
dfX = train.iloc[:,:-1]
dfy = train.iloc[:,[-1]]

In [23]:
# oversampling
def merge(df, count):
    if count >= 10: ###cover type 1,2를 얼마만큼 중복생성할 것인지 결정.### 
        return df
    add_1 = train.loc[train['Cover_Type'] == 1]
    add_2 = train.loc[train['Cover_Type'] == 2]
    new_df = pd.concat([add_1, add_2, df], axis=0)
    return merge(new_df, count+1)

new_train = merge(train, 0)

#데이터 분리
dfX = new_train.iloc[:, :-1]
dfy = new_train.iloc[:, [-1]]

# Cross Validation (K-fold)

In [24]:
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
#k_fold = KFold(n_splits=10, shuffle=True, random_state=0)

def cv_score(model, train_data, train_target):
    cv = KFold(n_splits=5, shuffle = True, random_state = 42)
    return cross_val_score(model, train_data, train_target, scoring='accuracy', cv= cv, n_jobs=-1)

# Data Split

In [25]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(dfX, dfy, test_size=0.3)

# Models

In [26]:
xgb_clf =  XGBClassifier(n_estimators=500, max_depth = 10, learning_rate=0.03 )
gboost_clf = GradientBoostingClassifier(n_estimators=200)
dt_clf = DecisionTreeClassifier(max_depth=10)
extra_clf = ExtraTreesClassifier(n_estimators=200, max_depth=10)
rf_clf = RandomForestClassifier(n_estimators=200, max_depth=10)

# models fitting

In [ ]:
%%time
xgb_clf.fit(X_train, y_train)
#gboost_clf.fit(X_train, y_train)
dt_clf.fit(X_train, y_train)
extra_clf.fit(X_train, y_train)
rf_clf.fit(X_train, y_train)

/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/preprocessing/label.py:95: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/sklearn/preprocessing/label.py:128: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


# Voting

In [ ]:
voting_clf = VotingClassifier(estimators=[('xgb', xgb_clf),('dt',dt_clf),
                                          ('extra', extra_clf),('rf',rf_clf)], voting='soft')

# Classification_report by each models

In [ ]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report

target = ['class1', 'class2','class3','class4','class5','class6','class7' ]

for clf in (xgb_clf, dt_clf, extra_clf, rf_clf, s_voting_clf):
    clf.fit(X_train, y_train)
    y_pred = clf.predict(X_test)
    print(clf.__class__.__name__, accuracy_score(y_test, y_pred))
    print (classification_report(y_test, y_pred, target_names=target))

# Submission

In [ ]:
final_clf = VotingClassifier(estimators=[('xgb', xgb_clf),('dt',dt_clf),
                                          ('extra', extra_clf),('rf',rf_clf)], voting='soft')
final_clf.fit(dfX, dfy)

test_data = test.drop("Id", axis=1).copy()
prediction = final_clf.predict(test_data)

In [ ]:
submission = pd.DataFrame({
        "Id": test["Id"],
        "Cover_type": prediction
    })

submission.to_csv('submission5.csv', index=False)

In [ ]:
submission = pd.read_csv('submission5.csv')
submission.head()

In [ ]:
submit_group = submission.groupby('Cover_type')
submit_group.count()